Add draft governance ZIPs (#989)
zcash/zips: zips/draft-ecc-zbloc.md

In cases where the zBloc governance processes could not be used for this reason, the consensus node implementors MUST, once the risk of exploitation has passed, publish a thorough post-mortem analysis that explains why this was the case.

## Funding

### Zcash Community Fund

A pool of multisig-controlled funds, seeded from the existing contents of the ZIP 1015 Deferred Dev Fund Lockbox [^zip-1015-lockbox] and supplemented with a funding stream consising of 20% of the block subsidy starting at block 3146400 that will continue until modified by a future ZIP, forms a new Zcash Community Fund. The mechanisms for the creation and management of this fund are described by the Deferred Dev Fund Lockbox Disbursement proposal [^draft-ecc-lockbox-disbursement], with the $\mathsf{stream\_value}$ parameter set to 20%, and the $\mathsf{stream\_end\_height}$ parameter set to the height at which the Zcash block subsidy diminishes to zero.

> The Zcash Community Fund will be controlled by a multisig with keys held by the Financial Privacy Foundation, the Zcash Foundation and the Electric Coin Company, along with two other entities yet to be determined ("Key-Holder Organisations").

Reviewer: Remove FPF

It is built to allow new innovative and positive voices to rise up and carry the flag of the Zcash community forward, in recognition of the risk that any one group may become captured or ineffective.

# Requirements

* This ZIP covers how to decide on Proposals for potential changes to Zcash Governance and/or disbursement of Protocol-defined Ecosystem Funding.
* There is a well-defined, publically agreed, process for evaluation and community feedback on both Governance Proposals and Grant Proposals.
* Funds are held in a multisig resistant to compromise of some of the parties' keys, up to a given threshold.
* The voting mechanism used for decision-making is similarly resistant to compromise of some of the parties' keys, up to a given threshold.
* No single party's non-cooperation or loss of keys is able to cause any Protocol-defined Ecosystem Funding to be locked up unrecoverably.
* 80% of the block rewards are given to miners, and 20% to a Zcash Community Fund seeded from the Deferred Dev Fund Lockbox.

> * The funds from the 20% funding stream will be usable immediately.

Reviewer 2: "immediately" is not well-defined; same issue as https://github.com/zcash/zips/pull/989/files#r2003460116

> * The funds in the Deferred Dev Fund Lockbox will be usable once a mechanism is available for disbursing them.

Reviewer 2: Should the disbursement mechanism options be factored out into a separate ZIP that both the governance ZIPs refer to? In both proposals, the model is effectively "grant-based funding, where development funds are controlled by a multisig" with different options for how and when funds may be disbursed to grantees.

Reviewer: Will someone take responsibility to draft this?

Author: @nuttycom agreed to do this.

* Any use of Deferred Dev Fund Lockbox funds is consistent with the purpose specified in [^zip-1015-lockbox] of "funding grants to ecosystem participants".

# Non-requirements

Any changes to the ZCG or its governance, such as its expansion to include more members, may be desired but are specifically outside this proposal’s scope.

> The full implementation of this proposal would require the design and implementation of a lockbox distribution mechanism, likely using multisig. This mechanism would be similar to that needed for an alternative proposal, the Community and Coin-holder Funding Model [^draft-ecc-community-and-coinholder]. However, this proposal does not require that a lockbox distribution mechanism be implemented at the same time as a change to the funding stream.

Reviewer 2: This gets to my comment above: any of the ecc-community-and-coinholder options would work for this proposal as well.

### Decision-making policies

To be considered a Constituency, each must pre-commit to an internal policy for decision-making that all other Constituencies approve of (i.e. greater than one decision maker, means of assessing community sentiment, etc.). This could be by polling, majority vote within the Constituency, coin-holder signaling, technical feasibility, mission alignment, etc.

In order for a Constituency to change its decision-making policy, the change MUST be expressed as a [Governance Proposal] and Approved before it comes into effect. The Constituency that is proposing to change its policy is eligible to vote, but it MUST make its decision according to its *pre-existing* policy (if it does not abstain).

## Initial Structure

The initial group of Constituencies will be (if they are willing and approved by the others and by the Zcash community):

* Coin-holders
* Electric Coin Company
* Zcash Foundation
* Shielded Labs
* Zcash Community Grants
* Zcash Core Engineers
* ZecHub

Each Constituency initially receives three Voting Units. This MAY be changed by a Governance Proposal.

> Constituencies other than the Coin-holder Constituency above are called Organizational Constituencies. The Organizational Constituencies will together elect a representative responsible for establishing a clear process and mechanism for gathering coin-holder sentiment, which will then be used as the decision-making process of the Coin-holder Constituency.

Reviewer 2: One possibility that this doesn't take into account is that an organic coin-holder constituency may choose to self-organize; the way that this is worded makes it sounds like the Organization Constituencies will choose the decision-making process instead of just the mechanism for determining coin-holder sentiment.
```suggestion
Constituencies other than the Coin-holder Constituency above are called Organizational Constituencies. The Organizational Constituencies will together elect a representative responsible for establishing a clear process and mechanism for gathering coin-holder sentiment, which will then be used to determine the will of the Coin-holder Constituency.
```

Each independent zBloc Constituency decides how to cast its Voting Units for each Governance or Funding Proposal, provided it is eligible to vote on that proposal (see the next section). If its decision-making policy permits, it MAY cast fewer than its total number of Voting Units, and MAY cast some Voting Units for Approval and some for Rejection. This might be used to publically acknowledge internal dispute over a decision, for example.

### Conflicts of interest

If a Constituency has a conflict of interest for a given decision, such as potentially being the recipient of funds or additional authority, then its Voting Units are not eligible to be cast for that vote. This affects the [Approval Threshold] discussed below, which is calculated as a proportion of the remaining Eligible Voting Units.

A Constituency is not automatically considered to have a conflict of interest merely because the decision affects it — for example, when changing its decision-making policy, or because it would need to make changes to software if the Proposal were Approved.

### Approval Thresholds

> The Approval Threshold is the absolute majority, as a proportion of the total number of Eligible Voting Units, required for a Proposal to be Approved. As stated earlier, Eligible Voting Units do not include those of Constituencies that have a conflict of interest for this Proposal. However, they do include the Voting Units of Constituencies that have no conflict of interest but choose to abstain.

Reviewer 2: We need something better than this: we need a system that makes it easy to choose between multiple possible alternatives. For example, consider the situation wherein ECC, ZF, Shielded Labs, and the Zingoistas all submit grant proposals for the same work; in this circumstance, is the grantee selected by determining which of the competing proposals has the highest number of approvals? What happens in the case of ties? 

In addition, timing constraints based upon when a grant proposal is made can introduce problems - for example, if a vote is scheduled for a grant, and then a few days before the vote a competing proposal is introduced, what happens? If a competing, but better proposal is introduced after the grant has been approved, what happens?

This comment also applies to the coinholder-and-community ZIP.

Reviewer 2: Note that if additional proposals are made beyond these first two, we're *already* having to face this problem. It will be very common - it will probably arise for literally every proposal.

Reviewer: It would be approved when the signing threshold is reached. At that point, it becomes a commitment that would only be overturned if a new threshold signals that it should be.

Reviewer 2: @jswihart the problem that I'm pointing out is that if there are multiple competing proposals, voting on each one individually doesn't work because you might end up in a situation where both grants are approved, but the community only wants to spend money on one of them and there's no way to signal that using the currently-described voting mechanism.

Author: I was originally thinking that negotiation about alternatives could be done off-chain, and then the on-chain vote would just be a ratification. If we want to put more of the sausage-making on-chain then we can, but I'm not sure that is a good idea, because choosing between multiple alternatives [fundamentally](https://en.wikipedia.org/wiki/Arrow%27s_impossibility_theorem) requires a more complicated voting mechanism.

> By default, a Governance Proposal requires a three-fifths Approval Threshold, and MUST have a Deadline at least two weeks from when the Proposal is published on-chain.

Reviewer 2: This seems out-of-order; we don't yet have any information about what it means to publish a Proposal on-chain, or who is responsible for that process.

Author: It was hard to find a sensible order that didn't involve forward-references. There might be a better order, but in any case remaining forward-references should be explicit links; feel free to fix that.

> All Consituencies MUST be consulted in order to determine whether they consider a draft Proposal to be "controversial", before it published on-chain. If *any* Constituency decides that a Proposal should be treated as controversial, it is then subject to a higher Approval Threshold of three quarters, and a Deadline at least four weeks from when the Proposal is published on-chain. (In the case of the Coin-holder constituency, a coin-holder vote is not necessarily required to establish this; it is likely sufficient to decide based on forum discussions for example.)

Reviewer 2: This strikes me as a can of worms. This means that an initial coinholder vote must be held to determine whether a proposal is "controversial" for every proposal.

Author: I see your point. The intent was that a coinholder vote would only need to be done to confirm that a proposal is "controversial" if a coalition of people who plausibly claim to collectively have significant holdings are arguing that it is.

The following categories of Governance Proposal are automatically subject to the three-quarters Approval Threshold and 4-week Deadline:

* Increasing or decreasing the voting power of zBloc Constituencies by adjusting how many Voting Units they hold.
* Increasing or decreasing the number of Funding Pools, the rules that govern them, and the amounts allocated.

In the case of Governance Proposals that affect consensus, final ratification occurs when node operators adopt the software implementing the changes.

> Note: Nothing forces developers of Zcash consensus node software to follow this specification. The aim of a process specification like this one is only to establish a social consensus. It fundamentally cannot affect the autonomy of developers of Zcash consensus node software to publish (or not) the software they want to publish, or the autonomy of node operators to run (or not) the software they want to run.

Reviewer 2: These facts are also true for the `community-and-coinholder` proposal (or, indeed, any proposal.) They should be copied there.

Author: True, but community-and-coinholder is only attempting to handle funding proposals, and so it's obvious in that case that it's not making any demands of node developers.

## Proposals

As stated in [Terminology], there are two categories of Proposal that can be put to a vote of the [zBloc Constituencies](#zbloc-membership):

* a "Governance Proposal", expressing a proposed change in Zcash Governance;
* a "Funding Proposal", expressing a proposed disbursement of Protocol-defined Ecosystem Funding.

Proposals have a Deadline (expressed as a Deadline Height if [^draft-ecc-onchain-accountable-voting] is used). If the Deadline is reached, they are implicitly Rejected. This does not necessarily preclude reintroducing the same or a similar Proposal.

### Governance Proposals

At least the following changes to Zcash Governance MUST be proposed and approved by a Governance Proposal before they can take effect:

* Changes to Zcash consensus rules. In this case Approval is REQUIRED in order for an activation height to be set.
* Substantive changes to the ZIP Process defined in ZIP 0 [^zip-0000].
* Substantive changes to this ZIP defining the zBloc.
* Changes to the set of zBloc Constituencies and/or the distribution of Voting Units among those Constituencies.

> * Changes to the policy a Constituency uses for its decision-making.

Reviewer 2: This last one should only apply to the coinholder constituency; it should not require a Governance Proposal for ECC to change its internal decision-making mechanism.

Author: Oh, I misinterpreted Josh's intent here then (given his thumbs-up).

### Disputes

If any zBloc Constituency, or the ZIP Editors, have concerns relating to the conduct of any vote, they MAY raise a dispute. Possible reasons to raise a dispute include believing that:

* A Constituency or any holders of keys for its Voting Units were unduly pressured or coerced.
* A Constituency voted contrary to its intended position (due to key compromise or error, for instance), or contrary to its stated decision-making policy.
* A Constituency failed to disclose a relevant conflict of interest.
* There were technical irregularities in the use of the voting mechanism.

If a dispute is raised, the Constituencies SHALL first resolve any technical issues as far as possible (e.g. by rotating keys), and then make one or more Governance Proposals intended to resolve the situation to the satisfaction of the community.

### Exceptional decisions to remediate security vulnerabilities

> The governance processes specified in this ZIP are not intended to prevent consensus node implementors from making consensus changes that they reasonably believe are critical to maintain the security of the Zcash Mainnet network and/or Zcash users. Such changes might need to be developed and released according to a controlled disclosure process that precludes public voting.

Reviewer 2: This is also useful to copy to the community-and-coinholder proposal, or to factor out into a shared preamble.

Author: Again, community-and-coinholder only handles grants, not any other aspect of governance.

In cases where the zBloc governance processes could not be used for this reason, the consensus node implementors MUST, once the risk of exploitation has passed, publish a thorough post-mortem analysis that explains why this was the case.

## Funding Proposals

### Zcash Community Fund

A funding stream will be established for a Zcash Community Fund, consisting of 20% of the block subsidy paid to a 2-of-3 P2SH multisig with keys held by the Financial Privacy Foundation, the Zcash Foundation and the Electric Coin Company ("Key-holder Organisations").

> This funding stream will start immediately on activation of this ZIP, which is assumed to be after the expiry of the existing ``FS_FPF_ZCG`` and ``FS_DEFERRED`` funding streams. It is proposed to be continued until modified by a future network upgrade.

Reviewer 2: Raising again that I don't like the funding mechanism to be coupled to the governance mechanism. Also, is the FPF intended to be included here?

What is the rationale behind the keyholders being different from the constituencies? I can see where the coinholder vote would need its disbursement power to be split among the other constituencies, but otherwise it's not obvious why we should have separate keyholders. Add a rationale if this is actually the chosen design.

* A decision to disburse funds is subject to veto if any Key-Holder Organization judges it to violate applicable law or reporting requirements.
* The Key-Holder Organizations SHALL be contractually required to recognize the Zcash Community Fund as a Restricted Fund donation or equivalent [TBD], and keep separate accounting of its balance and usage under its Transparency and Accountability obligations defined in ZIP 1015 [^zip-1015-transparency-and-accountability].

TODO: use the same veto process as [^draft-ecc-community-and-coinholder].

## Security precautions

The Key-Holder Organizations and the Organizational Consituencies MUST take appropriate precautions to safeguard funds from theft or accidental loss. Any theft or loss of funds, or any loss or compromise of key material MUST be reported to the Zcash community as promptly as possible after applying necessary mitigations.

# Acknowledgements

> Thank you to Tatyana Vinogradova, Kris Nuttycombe, and Chris Tomeo for the reviews, feedback, questions, and suggestions needed to make this a strong proposal.

Reviewer: ```suggestion
Thank you to Tatyana Peacemonger, Kris Nuttycombe, and Chris Tomeo for the reviews, feedback, questions, and suggestions needed to make this a strong proposal.
```
